### PR TITLE
Feature 953 spline to curvilinear

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGridSplineToGrid.hpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                           const UInt whichSpline,
+                           const DoubleVector& intersectionPoints,
+                           const double ssq) const;
+
+        /// @brief Get only the non-zero values from a row of the intersection matrix (paktij)
+        DoubleVector CompressRow(const VectorOfDoubleVectors& splineIntersections,
+                                 const UInt whichRow) const;
+
+        /// @brief Determine the intersection point or points of two splines
+        void DetermineIntersection(Splines& splines,
```
Comment (lucacarniato) [inline libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGridSplineToGrid.hpp:180]: const Splines&

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 2]
```diff
@@ -0,0 +1,904 @@
...
+
+    // Make copy of spline because the may be changed
+    // i.e. the order may change and the number of spline points may increase.
+    Splines splinesCopy(splines);
+    VectorOfDoubleVectors splineIntersections(splines.GetNumSplines(), DoubleVector(splines.GetNumSplines(), 0.0));
+
+    UInt mRefinement = curvilinearParameters.m_refinement;
+    UInt nRefinement = curvilinearParameters.n_refinement;
+
+    // mn12 in Fortran.
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:37]: remove fortran

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 3]
```diff
@@ -0,0 +1,904 @@
...
+    VectorOfThreeInts splineInteraction(splines.GetNumSplines());
+    UInt numNSplines = 0;
+
+    ComputeSplineIntersections(splinesCopy, splineIntersections, numNSplines);
+    OrderSplines(splinesCopy, numNSplines, splineIntersections);
+    DetermineSplineOrientation(splinesCopy, numNSplines, splineIntersections, splineInteraction);
+    GenerateGrid(splinesCopy, splineIntersections, splineInteraction, numNSplines, mRefinement, nRefinement, grid);
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:44]: can you return the grid instead? can we use the move constructor for the curvilinear grid?

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 4]
```diff
@@ -0,0 +1,904 @@
...
+
+    return grid;
+}
+
+bool meshkernel::CurvilinearGridSplineToGrid::CheckSplines(const Splines& splines) const
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:49]: rename to AreSplinesValid?

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 5]
```diff
@@ -0,0 +1,904 @@
...
+        firstNormalisedIntersectionLength = constants::missing::doubleValue;
+        crossProductOfIntersection = constants::missing::doubleValue;
+        numberTimesCrossing = 0;
+    }
+}
+
+void meshkernel::CurvilinearGridSplineToGrid::DoubleSplinePoints(Splines& splines) const
+{
+    std::vector<Point> doubledSplinePoints;
+    doubledSplinePoints.reserve(2 * splines.Size(splines.MaxSizeIndex()) - 1);
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:89]: 2 looks like a factor that might deserve its variable, faction can be made more generic by not specifying the Double in the name of the function?

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 6]
```diff
@@ -0,0 +1,904 @@
...
+    if (numberTimesCrossing == 1)
+    {
+
+        if (splineType[splineI] * splineType[splineJ] == 1)
+        {
+            if (maxSplineSize > MaximumNumberOfSplinePoints / 2)
+            {
+                throw AlgorithmError("splines both in m- and n-direction: spline {} and {}", splineI + 1, splineJ + 1);
+            }
+            else
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:137]: redundant else

        if (splineType[splineI] * splineType[splineJ] == 1)
        {
            if (maxSplineSize > MaximumNumberOfSplinePoints / 2)
            {
                throw AlgorithmError("splines both in m- and n-direction: spline {} and {}", splineI + 1, splineJ + 1);
            }
            return true;
        }
        if (splineType[splineI] == 0 && splineType[splineJ] == 0)
        {
            throw AlgorithmError("Both spline {} and {} are undefined", splineI + 1, splineJ + 1);
        }

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 7]
```diff
@@ -0,0 +1,904 @@
...
+            if (splineIntersections[j][i] != 0.0)
+            {
+                maxn = std::max(maxn, splineInteraction[j][1]);
+            }
+        }
+
+        splineInteraction[i][0] = maxn;
+    }
+
+    // Dan alles ranken in M richting
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:387]: remove dutch comments

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 8]
```diff
@@ -0,0 +1,904 @@
...
+    for (UInt i = numNSplines; i < splines.GetNumSplines(); ++i)
+    {
+        int maxm = 0;
+
+        for (UInt j = 0; j < numNSplines; ++j)
+        {
+            maxm = 0;
+            UInt iilast = numNSplines;
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:395]: use meaningful variable names

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 9]
```diff
@@ -0,0 +1,904 @@
...
+                                                                          double& normalisedDistance,
+                                                                          double& intervalLength) const
+{
+    Point startPoint = splines.m_splineNodes[whichSpline][0];
+    Point endPoint;
+    double stepSize = 0.1;
+    double intervalStart = 0.0;
+    double intervalEnd;
+
+    // in getdis.f90 this is: normalisedDistance = std::min(normalisedDistance, static_cast<double>(splines.m_splineNodes[whichSpline].size()))
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:522]: also here remove comments

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp [hunk 10]
```diff
@@ -0,0 +1,904 @@
...
+}
+
+meshkernel::Point meshkernel::CurvilinearGridSplineToGrid::ComputePoint(const Splines& splines,
+                                                                        const UInt whichSpline,
+                                                                        const DoubleVector& intersectionPoints,
+                                                                        const double ssq) const
+{
+    Point result;
+
+    double ax = intersectionPoints[0];
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:626]: const double ax = intersectionPoints[0];
    const double cx = intersectionPoints[intersectionPoints.size() - 1];

    FuncAdimensionalToDimensionalDistanceOnSpline func(splines, whichSpline, false, 0.0, ssq);
    const double distance = FindFunctionRootWithGoldenSectionSearch(func, ax, cx);

    return splines.Evaluate(whichSpline, distance);

**Review by lucacarniato — CHANGES_REQUESTED** [review]

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp (hunk 5 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:89]: This is possible, but for a more general increase factor, will considerably increase the complexity of this function.

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp (hunk 2 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:37]: done

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp (hunk 4 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:49]: done

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp (hunk 8 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:395]: done

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp (hunk 5 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:89]: Changed, with comment on constant that the function IncreaseSplinePoints has to be changed to reflect the different scale factor.

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGridSplineToGrid.hpp (hunk 1 above)
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGridSplineToGrid.hpp:180]: done

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp (hunk 3 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:44]: It should already use the move constructor

Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSplineToGrid.cpp:43]: Add addition function to mode grid nodes

**Review by lucacarniato — APPROVED** [review]

**Pull request merged**